### ISSUE TITLE
Tuple types get incorrect contextual type

### ISSUE THREAD

**Issue by Taytay** [Microsoft/TypeScript#9901]
**TypeScript Version:** 
```
$ tsc --version
Version 2.0.0
```

**Code**

```ts
// concat-bug.ts
var a : Array<[number, number]>= [[1, 2]];

// Typescript detects these first two tuples as arrays of numbers (`number[]`) instead of `[number, number]`
a.concat([[3, 4], [5, 6], [7, 8]]);
```

**Expected behavior:**
No compile errors. (It works in the 1.8 playground)

**Actual behavior:** 
```
$ tsc concat-bug.ts
concat-bug.ts(4,10): error TS2345: Argument of type '[number[], number[], [number, number]]' is not assignable to parameter of type '[number, number] | [number, number][]'.
  Type '[number[], number[], [number, number]]' is not assignable to type '[number, number][]'.
    Types of property 'pop' are incompatible.
      Type '() => number[] | [number, number]' is not assignable to type '() => [number, number]'.
        Type 'number[] | [number, number]' is not assignable to type '[number, number]'.
          Type 'number[]' is not assignable to type '[number, number]'.
            Property '0' is missing in type 'number[]'.
```

Note that the 3rd parameter is correctly detected as a tuple, but the first two are detected as arrays of numbers.

**Comment by Taytay** [comment]
Thanks @mhegazy

**Comment by sandersn** [comment]
This is a bug caused by removing one of the overloads of `concat` (the ugly one, with an additional, constrained type parameter). At the time we thought it wouldn't break anything. I'll try to find the discussion from that time, but almost certainly the right fix is to put the other overload back.

**Comment by sandersn** [comment]
PR is up at #9997 -- I added an explanation of why this fails there.

**Comment by Taytay** [comment]
TY for the fix!
